modules/wallet/defrag.go

```diff
@@ -0,0 +1,51 @@
+package wallet
+
+import (
+	"sort"
+
+	"github.com/NebulousLabs/Sia/types"
+)
+
+const (
+	// defragThreshold is the number of outputs a wallet is allowed before it is
+	// defragmented.
+	defragThreshold = 20
+)
+
+// defragWallet computes the sum of the 15 largest outputs in the wallet and
+// sends that sum to itself, effectively defragmenting the wallet. This defrag
+// operation is only performed if the wallet has greater than defragThreshold
+// outputs.
+func (w *Wallet) defragWallet() {
```

Comment:
In the contractor I use a `TryMutex` to accomplish this: https://github.com/NebulousLabs/Sia/blob/master/modules/renter/contractor/update.go#L66
Not sure if it's the best pattern but it seems to work.